lib/spack/spack/ci/common.py

```diff
@@ -0,0 +1,826 @@
+        self.check_index_only = check_index_only
+        self.broken_specs_url = broken_specs_url
+        self.rebuild_index = rebuild_index
+        self.untouched_pruning_dependent_depth = untouched_pruning_dependent_depth
+        self.prune_untouched = prune_untouched
+        self.prune_up_to_date = prune_up_to_date
+        self.prune_external = prune_external
+        self.stack_name = stack_name
+        self.pipeline_type = pipeline_type
+        self.require_signing = require_signing
+        self.cdash_handler = cdash_handler
+
+
+class PipelineNode:
+    spec: spack.spec.Spec
+    parents: Set[str]
+    children: Set[str]
+
+    def __init__(self, spec: spack.spec.Spec):
+        self.spec = spec
+        self.parents = set()
+        self.children = set()
+
+
+class PipelineDag:
+    """Turn a list of specs into a simple directed graph, that doesn't keep track
+    of edge types."""
+
+    @classmethod
+    def key(cls, spec: spack.spec.Spec) -> str:
+        return spec.dag_hash()
+
+    def __init__(self, specs: List[spack.spec.Spec]) -> None:
+        # Build dictionary of nodes
+        self.nodes: Dict[str, PipelineNode] = {
+            PipelineDag.key(s): PipelineNode(s)
+            for s in traverse.traverse_nodes(specs, deptype=dt.ALL_TYPES, root=True)
+        }
+
+        # Create edges
+        for edge in traverse.traverse_edges(
+            specs, deptype=dt.ALL_TYPES, root=False, cover="edges"
+        ):
+            parent_key = PipelineDag.key(edge.parent)
+            child_key = PipelineDag.key(edge.spec)
+
+            self.nodes[parent_key].children.add(child_key)
+            self.nodes[child_key].parents.add(parent_key)
+
+    def prune(self, node_key: str):
+        """Remove a node from the graph, and reconnect its parents and children"""
+        node = self.nodes[node_key]
+        for parent in node.parents:
+            self.nodes[parent].children.remove(node_key)
+            self.nodes[parent].children |= node.children
+        for child in node.children:
+            self.nodes[child].parents.remove(node_key)
+            self.nodes[child].parents |= node.parents
+        del self.nodes[node_key]
+
+    def traverse(self, top_down: bool = True):
+        """If top_down is True, yields the roots first, followed by direct children
+        of roots, followed by their children and so on.  Otherwise yields the leaves
+        first, followed by nodes whose only children were leaves, etc."""
+        if top_down:
+            get_in_edges = lambda key: sorted([key for key in self.nodes[key].parents])
+            get_out_edges = lambda key: sorted([key for key in self.nodes[key].children])
+        else:
+            get_in_edges = lambda key: sorted([key for key in self.nodes[key].children])
+            get_out_edges = lambda key: sorted([key for key in self.nodes[key].parents])
+
+        # Create a representation of the DAG we can consume as we traverse
+        entire_graph = {
```

Comment:
This is a rather non-standard data structure for a graph. Usually people have `nodes` / `edges` and that's it. I don't like the abundance of `_`s below as a result of this choice.

You do not need to collect and sort in-edges, you just have to know their count. Can you work with:

```
out_edges = {key: sorted(...) ...}
num_in_edges = {key: len(...) ...}
```

instead?

Reply:
Notice that we have a different ordering elsewhere `spack.traverse.sort_edges`. I think it's good to sort by name so your tests are robust when hashes change for whatever reason.